### ISSUE TITLE
Abbreviating Map.sel with op_String_Access fails to verify

### ISSUE THREAD

**Issue by aseemr** [FStarLang/FStar#1226]
```
module Test

#set-options "--initial_fuel 2 --max_fuel 2 --initial_ifuel 1 --max_ifuel 1"

unfold let op_String_Access = Map.sel

let test (m : Map.t int int) (addr : int) : Lemma
  (requires (m.[addr] = 2))
  (ensures  (True))
  = assert (m.[addr] = 2);
    ()
```

Using `Map.sel m addr` in the `assert` in the lemma body works.

Removing `()` from the lemma body works.

Defining `op_String_Access` like: `unfold let op_String_Access (#key:eqtype) (#value:Type) (m:Map.t key value) (k:key)     = Map.sel m k` works.

**Comment by aseemr** [comment]
Also, removing `unfold` qualifier from `op_String_Access` works.

**Comment by nikswamy** [comment]
Here's a workaround. It's a bit bizarre why it needs the annotations ... I can dig more.

```module Bug1226
#set-options "--initial_fuel 2 --max_fuel 2 --initial_ifuel 1 --max_ifuel 1"

unfold let op_String_Access (#a:eqtype) (#b:Type) (x:Map.t a b) (y:a) : Tot b = Map.sel x y

let test (m : Map.t int int) (addr : int) : Lemma
  (requires (m.[addr] = 2))
  (ensures  (True))
  = assert (m.[addr] = 2);
    ()
```

**Comment by nikswamy** [comment]
Some more investigation ... seems to be related to some weirdness with the way annotations on `val` vs. annotations on the corresponding `let` are treated. For instance, this works fine:

```
module Bug1226
unfold let op_String_Access = Map.sel

val test (m : Map.t int int) (addr : int) : Lemma
  (requires (m.[addr] == 2))
  (ensures  (True))

let test m addr = assert (m.[addr] == 2); ()
```

**Comment by nikswamy** [comment]
Should be fixed by 1f9bd88 ... awaiting some feedback on that commit before pushing to master.

What was happening was that the inferred definition of `op_String_Access` was being tagged with an internal annotation that was needlessly blocking reduction, causing the generated VC to contain way more higher-order stuff than it ought to ... which eventually lead to Z3 being unable to prove this simple property.
